In [1]:
!pwd

/Users/shihosato/src/github.com/twinkle13531/master_degree/202012/Summary


In [2]:
%load_ext autoreload
%autoreload 2

In [3]:
import sys, os
sys.path.append(os.path.join('..', 'Random'))
sys.path.append(os.path.join('..', 'SA'))
sys.path.append(os.path.join('..', 'QA'))

In [4]:
import QA_DwaveSampler_AutoEmbedding as qda
import random_exact_test_functions as ret
import Neal_exact_test_functions as net
import pandas as pd
import numpy as np
import timeit
import matplotlib.pyplot as plt
import statistics

## 辞書の作成
### QA  {bit:{num:{chain_strength:[valid_y_num, total_y_num, mean_appearance_num, mean_broken_chain_percent]}}}

In [5]:
n, num_reads = 100, 100
for bit in [20, 25, 30]:
    for num in [0, 1, 2, 3, 4]:
        path = '../input/ost{}_{}.csv'.format(bit, num)
        df = pd.read_csv(path, sep=',')
        bqm = qda.make_Hamiltonian(df)
        for chain_strength in [10, 50 ,100, 500, 1000]:
            QDA = qda.QA_DSampler_AEmbedding(df, bqm, n, num_reads, chain_strength)
            valid_y_info_dic, calculation_time = QDA.find_valid_y_time_ntimes()
            print('when data = {}_{} and chain_strength={} number of valid y is {}'.format(bit, num, chain_strength, len(valid_y_info_dic)))

when data = 20_0 and chain_strength=10 number of valid y is 755
when data = 20_0 and chain_strength=50 number of valid y is 290
when data = 20_0 and chain_strength=100 number of valid y is 218
when data = 20_0 and chain_strength=500 number of valid y is 171
when data = 20_0 and chain_strength=1000 number of valid y is 145
when data = 20_1 and chain_strength=10 number of valid y is 460
when data = 20_1 and chain_strength=50 number of valid y is 93
when data = 20_1 and chain_strength=100 number of valid y is 51
when data = 20_1 and chain_strength=500 number of valid y is 43
when data = 20_1 and chain_strength=1000 number of valid y is 24


KeyboardInterrupt: embedding cancelled by keyboard interrupt

- when data = 20_0 and chain_strength=10 number of valid y is 755
- when data = 20_0 and chain_strength=50 number of valid y is 290
- when data = 20_0 and chain_strength=100 number of valid y is 218
- when data = 20_0 and chain_strength=500 number of valid y is 171
- when data = 20_0 and chain_strength=1000 number of valid y is 145
- when data = 20_1 and chain_strength=10 number of valid y is 460
- when data = 20_1 and chain_strength=50 number of valid y is 93
- when data = 20_1 and chain_strength=100 number of valid y is 51
- when data = 20_1 and chain_strength=500 number of valid y is 43
- when data = 20_1 and chain_strength=1000 number of valid y is 24

- when data = 30_0 and chain_strength=5 number of valid y is 4 and chain_break_% = 0.4
- when data = 30_0 and chain_strength=10 number of valid y is 11 and chain_break_% = 0.0696969696969697
- when data = 30_0 and chain_strength=50 number of valid y is 3 and chain_break_% = 0.0
- when data = 30_0 and chain_strength=100 number of valid y is 3 and chain_break_% = 0.0

In [12]:
n, num_reads = 100, 100
for bit in [40, 35, 30, 25, 20]:
    for num in [0, 1, 2, 3, 4]:
        path = '../input/ost{}_{}.csv'.format(bit, num)
        df = pd.read_csv(path, sep=',')
        bqm = qda.make_Hamiltonian(df)
        for chain_strength in [1, 5, 10, 50 ,100]:
            QDA = qda.QA_DSampler_AEmbedding(df, bqm, n, num_reads, chain_strength)
            valid_y_info_dic, calculation_time = QDA.find_valid_y_time_ntimes()
            broken_chain_mean_percent = statistics.mean([d[1] for d in valid_y_info_dic.values()])
            print('when data = {}_{} and chain_strength={} number of valid y is {} and chain_break_% = {}'.format(bit, num, chain_strength, len(valid_y_info_dic), broken_chain_mean_percent))

when data = 30_0 and chain_strength=5 number of valid y is 4 and chain_break_% = 0.4
when data = 30_0 and chain_strength=10 number of valid y is 11 and chain_break_% = 0.0696969696969697
when data = 30_0 and chain_strength=50 number of valid y is 3 and chain_break_% = 0.0
when data = 30_0 and chain_strength=100 number of valid y is 3 and chain_break_% = 0.0


StatisticsError: mean requires at least one data point

In [13]:
n, num_reads = 100, 100

QA_valid_y_info_dic = {}
for bit in [20, 25, 30]:
    
    QA_valid_y_info_dic[bit] = {}
    #print(QA_valid_y_info_dic)
    for num in [0, 1, 2, 3, 4]:
        QA_valid_y_info_dic[bit][num] = {}
        #print(QA_valid_y_info_dic)
        path = '../input/ost{}_{}.csv'.format(bit, num)
        df = pd.read_csv(path, sep=',')
        bqm = qda.make_Hamiltonian(df)

        for chain_strength in range(10, 28):
            QA_valid_y_info_dic[bit][num][chain_strength] = []
            #print(QA_valid_y_info_dic)
            QDA = qda.QA_DSampler_AEmbedding(df, bqm, n, num_reads, chain_strength)

            #valid_y_info_dic={sample:[occurrences, chain_break_fraction]}
            valid_y_info_dic, calculation_time = QDA.find_valid_y_time_ntimes()
            QA_valid_y_info_dic[bit][num][chain_strength].append(len(valid_y_info_dic))

            occurance_list = [int(valid_y_info[0]) for valid_y_info in list(valid_y_info_dic.values())]
            QA_valid_y_info_dic[bit][num][chain_strength].append(sum(occurance_list))#
            
            if sum(occurance_list)==0:
                QA_valid_y_info_dic[bit][num][chain_strength].append(0)#
                QA_valid_y_info_dic[bit][num][chain_strength].append(0)#
            else:
                QA_valid_y_info_dic[bit][num][chain_strength].append(statistics.mean(occurance_list))#
                broken_chain_percent_list = [int(valid_y_info[3]) for valid_y_info in valid_y_info_dic]
                QA_valid_y_info_dic[bit][num][chain_strength].append(statistics.mean(broken_chain_percent_list))#

OSError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))

In [ ]:
QA_valid_y_info_dic#n=100, num_reads=100のとき必ず100回毎回重複する

In [98]:
n, num_reads = 10, 1000

QA_valid_y_info_dic = {}
for bit in [20, 25, 30]:
    
    QA_valid_y_info_dic[bit] = {}
    #print(QA_valid_y_info_dic)
    for num in [0, 1, 2, 3, 4]:
        QA_valid_y_info_dic[bit][num] = {}
        #print(QA_valid_y_info_dic)
        path = '../input/ost{}_{}.csv'.format(bit, num)
        df = pd.read_csv(path, sep=',')
        bqm = qda.make_Hamiltonian(df)

        for chain_strength in range(10, 28):
            QA_valid_y_info_dic[bit][num][chain_strength] = []
            #print(QA_valid_y_info_dic)
            QDA = qda.QA_DSampler_AEmbedding(df, bqm, n, num_reads, chain_strength)

            #valid_y_info_dic={sample:[occurrences, chain_break_fraction]}
            valid_y_info_dic, calculation_time = QDA.find_valid_y_time_ntimes()
            QA_valid_y_info_dic[bit][num][chain_strength].append(len(valid_y_info_dic))

            occurance_list = [int(valid_y_info[0]) for valid_y_info in list(valid_y_info_dic.values())]
            QA_valid_y_info_dic[bit][num][chain_strength].append(sum(occurance_list))#
            
            if sum(occurance_list)==0:
                QA_valid_y_info_dic[bit][num][chain_strength].append(0)#
                QA_valid_y_info_dic[bit][num][chain_strength].append(0)#
            else:
                QA_valid_y_info_dic[bit][num][chain_strength].append(statistics.mean(occurance_list))#
                broken_chain_percent_list = [int(valid_y_info[3]) for valid_y_info in valid_y_info_dic]
                QA_valid_y_info_dic[bit][num][chain_strength].append(statistics.mean(broken_chain_percent_list))#

In [ ]:
bit = 20
_20bit_chain_maxnum = [                                                                                                                                                                                                               ]
for num in [0, 1, 2, 3, 4]:
    .append(QA_valid_y_info_dci[bit][num][10][0])

In [99]:
QA_valid_y_info_dic

{20: {0: {10: [66, 1240, 18.78787878787879, 0.09090909090909091],
   11: [49, 490, 10, 0.5918367346938775],
   12: [76, 770, 10.131578947368421, 0.3157894736842105],
   13: [78, 790, 10.128205128205128, 0.46153846153846156],
   14: [53, 530, 10, 0.2830188679245283],
   15: [60, 600, 10, 0.45],
   16: [65, 670, 10.307692307692308, 0.4307692307692308],
   17: [58, 590, 10.172413793103448, 0.3793103448275862],
   18: [79, 790, 10, 0.45569620253164556],
   19: [56, 560, 10, 0.14285714285714285],
   20: [63, 630, 10, 0.6349206349206349],
   21: [41, 410, 10, 0.7804878048780488],
   22: [43, 430, 10, 0.627906976744186],
   23: [56, 580, 10.357142857142858, 0.4107142857142857],
   24: [31, 310, 10, 0.2903225806451613],
   25: [46, 460, 10, 0.2391304347826087],
   26: [57, 570, 10, 0.45614035087719296],
   27: [47, 470, 10, 0.5106382978723404]},
  1: {10: [44, 460, 10.454545454545455, 0.7727272727272727],
   11: [50, 530, 10.6, 0.52],
   12: [29, 290, 10, 0.5862068965517241],
   13: [19, 200, 

In [102]:
n, num_reads = 1, 10000
QA_valid_y_info_dic = {}
for bit in [20, 25, 30]:
    
    QA_valid_y_info_dic[bit] = {}
    #print(QA_valid_y_info_dic)
    for num in [0, 1, 2, 3, 4]:
        QA_valid_y_info_dic[bit][num] = {}
        #print(QA_valid_y_info_dic)
        path = '../input/ost{}_{}.csv'.format(bit, num)
        df = pd.read_csv(path, sep=',')
        bqm = qda.make_Hamiltonian(df)

        for chain_strength in range(10, 28):
            QA_valid_y_info_dic[bit][num][chain_strength] = []
            #print(QA_valid_y_info_dic)
            QDA = qda.QA_DSampler_AEmbedding(df, bqm, n, num_reads, chain_strength)

            #valid_y_info_dic={sample:[occurrences, chain_break_fraction]}
            valid_y_info_dic, calculation_time = QDA.find_valid_y_time_ntimes()
            QA_valid_y_info_dic[bit][num][chain_strength].append(len(valid_y_info_dic))

            occurance_list = [int(valid_y_info[0]) for valid_y_info in list(valid_y_info_dic.values())]
            QA_valid_y_info_dic[bit][num][chain_strength].append(sum(occurance_list))#
            
            if sum(occurance_list)==0:
                QA_valid_y_info_dic[bit][num][chain_strength].append(0)#
                QA_valid_y_info_dic[bit][num][chain_strength].append(0)#
            else:
                QA_valid_y_info_dic[bit][num][chain_strength].append(statistics.mean(occurance_list))#
                broken_chain_percent_list = [int(valid_y_info[3]) for valid_y_info in valid_y_info_dic]
                QA_valid_y_info_dic[bit][num][chain_strength].append(statistics.mean(broken_chain_percent_list))#

In [103]:
QA_valid_y_info_dic

[autoreload of QA_DwaveSampler_AutoEmbedding failed: Traceback (most recent call last):
  File "/Users/shihosato/.pyenv/versions/miniconda3-4.7.12/lib/python3.7/site-packages/IPython/extensions/autoreload.py", line 245, in check
    superreload(m, reload, self.old_objects)
  File "/Users/shihosato/.pyenv/versions/miniconda3-4.7.12/lib/python3.7/site-packages/IPython/extensions/autoreload.py", line 394, in superreload
    module = reload(module)
  File "/Users/shihosato/.pyenv/versions/miniconda3-4.7.12/lib/python3.7/imp.py", line 314, in reload
    return importlib.reload(module)
  File "/Users/shihosato/.pyenv/versions/miniconda3-4.7.12/lib/python3.7/importlib/__init__.py", line 169, in reload
    _bootstrap._exec(spec, module)
  File "<frozen importlib._bootstrap>", line 630, in _exec
  File "<frozen importlib._bootstrap_external>", line 724, in exec_module
  File "<frozen importlib._bootstrap_external>", line 860, in get_code
  File "<frozen importlib._bootstrap_external>", line 791

{20: {0: {10: [579, 859, 1.4835924006908463, 0.7046632124352331],
   11: [726, 840, 1.1570247933884297, 0.3209366391184573],
   12: [782, 991, 1.2672634271099745, 0.4296675191815857],
   13: [622, 726, 1.167202572347267, 0.5192926045016077],
   14: [580, 645, 1.1120689655172413, 0.3017241379310345],
   15: [657, 716, 1.0898021308980212, 0.5616438356164384],
   16: [647, 679, 1.0494590417310665, 0.34775888717156106],
   17: [604, 631, 1.044701986754967, 0.4105960264900662],
   18: [474, 493, 1.040084388185654, 0.32489451476793246],
   19: [515, 539, 1.0466019417475727, 0.4174757281553398],
   20: [361, 371, 1.0277008310249307, 0.6509695290858726],
   21: [463, 486, 1.0496760259179265, 0.4017278617710583],
   22: [402, 416, 1.0348258706467661, 0.4129353233830846],
   23: [308, 316, 1.025974025974026, 0.5487012987012987],
   24: [542, 576, 1.062730627306273, 0.31365313653136534],
   25: [433, 447, 1.0323325635103926, 0.42725173210161665],
   26: [593, 609, 1.026981450252951, 0.35244519392

### random {bit:{num:[valid_y_num, total_y_num, mean_appearance_num]}}

In [92]:
num_reads = 10000
random_valid_y_info_dic = {}
for bit in [20, 25, 30, 35, 40]:
    
    random_valid_y_info_dic[bit] = {}
    for num in [0, 1, 2, 3, 4]:
        random_valid_y_info_dic[bit][num] = []
        df_path = '../input/ost{}_{}.csv'.format(bit, num)
        df = pd.read_csv(df_path, sep=',')
        t_list = ret.calc_marginals(df)
        valid_y_dic, calculation_time = ret.find_valid_y(df, num_reads)
        random_valid_y_info_dic[bit][num].append(len(valid_y_dic))

        occurance_list = list(valid_y_dic.values())
        random_valid_y_info_dic[bit][num].append(sum(occurance_list))
        if sum(occurance_list)==0:
            random_valid_y_info_dic[bit][num].append(0)
        else:
            random_valid_y_info_dic[bit][num].append(statistics.mean(occurance_list))

In [93]:
random_valid_y_info_dic

{20: {0: [159, 161, 1.0125786163522013],
  1: [32, 34, 1.0625],
  2: [10, 10, 1],
  3: [26, 26, 1],
  4: [33, 33, 1]},
 25: {0: [1, 1, 1], 1: [3, 3, 1], 2: [27, 27, 1], 3: [3, 3, 1], 4: [1, 1, 1]},
 30: {0: [1, 1, 1], 1: [9, 9, 1], 2: [1, 1, 1], 3: [2, 2, 1], 4: [0, 0, 0]},
 35: {0: [0, 0, 0], 1: [2, 2, 1], 2: [3, 3, 1], 3: [0, 0, 0], 4: [0, 0, 0]},
 40: {0: [0, 0, 0], 1: [1, 1, 1], 2: [1, 1, 1], 3: [0, 0, 0], 4: [0, 0, 0]}}

In [47]:
num_reads = 10000
SA_valid_y_info_dic = {}
for bit in [20, 25, 30, 35, 40]:
    
    SA_valid_y_info_dic[bit] = {}
    for num in [0, 1, 2, 3, 4]:
        SA_valid_y_info_dic[bit][num] = []
        path = '../input/ost{}_{}.csv'.format(bit, num)
        df = pd.read_csv(path, sep=',')
        res = net.make_res_data(df, num_reads)
        valid_y_info_dic = net.find_valid_y(res)
        
        SA_valid_y_info_dic[bit][num].append(len(valid_y_info_dic))

        occurance_list = list(valid_y_info_dic.values())
        SA_valid_y_info_dic[bit][num].append(sum(occurance_list))
            
        if sum(occurance_list)==0:
            SA_valid_y_info_dic[bit][num].append(0)
        else:
            SA_valid_y_info_dic[bit][num].append(statistics.mean(occurance_list))

In [48]:
SA_valid_y_info_dic

{20: {0: [7343, 9766, 1],
  1: [2832, 9826, 3],
  2: [1050, 9967, 9],
  3: [2580, 9856, 3],
  4: [2411, 9712, 4]},
 25: {0: [6764, 9843, 1],
  1: [4504, 8654, 1],
  2: [9545, 9973, 1],
  3: [7860, 9572, 1],
  4: [1259, 8455, 6]},
 30: {0: [8967, 9780, 1],
  1: [8822, 8863, 1],
  2: [9502, 9747, 1],
  3: [9265, 9553, 1],
  4: [8847, 9363, 1]},
 35: {0: [9530, 9570, 1],
  1: [9729, 9732, 1],
  2: [9692, 9697, 1],
  3: [9609, 9618, 1],
  4: [9161, 9206, 1]},
 40: {0: [9676, 9676, 1],
  1: [9664, 9664, 1],
  2: [9218, 9219, 1],
  3: [9895, 9895, 1],
  4: [8831, 8840, 1]}}